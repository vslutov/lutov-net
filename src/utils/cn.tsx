import React from "react"

export const appendCN = (...cns: Array<string | undefined>): string =>
  cns.filter((s) => s != null && s !== "").join(" ")

const cn = (tag: any, baseClassName: string, initialProps?: any) => {
  const WithClass = (
    { as, className, children, ...props }:
    { as?: any, className?: string, children?: React.ReactNode } & any
  ) => {
    const Tag = as != null ? as : tag
    className = appendCN(baseClassName, className)

    return <Tag {...initialProps} className={className} {...props}>{children}</Tag>
  }

  if (typeof tag === "string") {
    WithClass.displayName = `${tag} with class ${baseClassName}`
  } else if (tag?.displayName != null) {
    WithClass.displayName = `${tag.displayName} with class ${baseClassName}`
  }

  WithClass.cn = (Tag: any, appendClassName?: string) =>
    ({ className, ...props }: { className?: string} & any
    ) => {
      className = appendCN(appendClassName, className)
      return WithClass({ as: Tag, className, ...props })
    }

  return WithClass
}

export default cn
