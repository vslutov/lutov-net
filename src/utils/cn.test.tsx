/* eslint-env jest */

import React from "react"
import { render } from "enzyme"
import cn from "./cn"

describe("cn", () => {
  const MyDiv = (props: any) => <div {...props} />
  MyDiv.displayName = "MyDiv"

  const Span = cn("span", "MyClass")
  const Link = Span.cn("a")

  it("should render", () => {
    expect(render(<Span />)).toMatchSnapshot()
    expect(render(<Span className='addtional' />)).toMatchSnapshot()
    expect(render(<Span>hello</Span>)).toMatchSnapshot()
    expect(render(<Span as={MyDiv} />)).toMatchSnapshot()
    expect(render(<Span as='a' href='kek' />)).toMatchSnapshot()
  })

  it("should render as cn", () => {
    expect(render(<Link />)).toMatchSnapshot()
    expect(render(<Link className='addtional' />)).toMatchSnapshot()
    expect(render(<Link href='kek'>hello</Link>)).toMatchSnapshot()
    expect(render(<Link as='b' />)).toMatchSnapshot()
    expect(render(<Link as={MyDiv} />)).toMatchSnapshot()
  })

  it("should add classname on cn", () => {
    const LeftLink = Span.cn("a", "left")
    expect(render(<LeftLink />)).toMatchSnapshot()
  })

  it("should get display name from component", () => {
    const DivWithClass = cn(MyDiv, "MyDiv")
    expect(render(<DivWithClass />)).toMatchSnapshot()
  })
})
