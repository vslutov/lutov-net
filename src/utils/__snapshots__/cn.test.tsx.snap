// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cn should add classname on cn 1`] = `
<a
  class="MyClass left"
/>
`;

exports[`cn should get display name from component 1`] = `
<div
  class="MyDiv"
/>
`;

exports[`cn should render 1`] = `
<span
  class="MyClass"
/>
`;

exports[`cn should render 2`] = `
<span
  class="MyClass addtional"
/>
`;

exports[`cn should render 3`] = `
<span
  class="MyClass"
>
  hello
</span>
`;

exports[`cn should render 4`] = `
<div
  class="MyClass"
/>
`;

exports[`cn should render 5`] = `
<a
  class="MyClass"
  href="kek"
/>
`;

exports[`cn should render as cn 1`] = `
<a
  class="MyClass"
/>
`;

exports[`cn should render as cn 2`] = `
<a
  class="MyClass addtional"
/>
`;

exports[`cn should render as cn 3`] = `
<a
  class="MyClass"
  href="kek"
>
  hello
</a>
`;

exports[`cn should render as cn 4`] = `
<b
  class="MyClass"
/>
`;

exports[`cn should render as cn 5`] = `
<div
  class="MyClass"
/>
`;
